### ISSUE TITLE
Fail to build for ppc bigendian

### ISSUE THREAD

**Issue by marillat** [gpac/gpac#1511]
powerpc Debian unstable, gcc  9.3.0 (Debian 9.3.0-13)

````
evg/raster_yuv.c: In function ‘overmask_yuv420p_10’:
evg/raster_yuv.c:1294:55: error: ‘ptr’ undeclared (first use in this function)
 1294 | #define get_u16_le(val, _ptr) { val = ((u32) (*(u8 *) ptr+1)<< 8) | *(u8 *) ptr; }
      |                                                       ^~~
evg/raster_yuv.c:1329:2: note: in expansion of macro ‘get_u16_le’
 1329 |  get_u16_le(dstc, dst);
      |  ^~~~~~~~~~
evg/raster_yuv.c:1294:55: note: each undeclared identifier is reported only once for each function it appears in
 1294 | #define get_u16_le(val, _ptr) { val = ((u32) (*(u8 *) ptr+1)<< 8) | *(u8 *) ptr; }
      |                                                       ^~~
evg/raster_yuv.c:1329:2: note: in expansion of macro ‘get_u16_le’
 1329 |  get_u16_le(dstc, dst);
      |  ^~~~~~~~~~
evg/raster_yuv.c: In function ‘evg_yuv420p_10_flush_uv_const’:
evg/raster_yuv.c:1294:55: error: ‘ptr’ undeclared (first use in this function)
 1294 | #define get_u16_le(val, _ptr) { val = ((u32) (*(u8 *) ptr+1)<< 8) | *(u8 *) ptr; }
      |                                                       ^~~
evg/raster_yuv.c:1376:5: note: in expansion of macro ‘get_u16_le’
 1376 |     get_u16_le(dst, s_ptr_u);
      |     ^~~~~~~~~~
evg/raster_yuv.c: In function ‘evg_yuv420p_10_flush_uv_var’:
evg/raster_yuv.c:1294:55: error: ‘ptr’ undeclared (first use in this function)
 1294 | #define get_u16_le(val, _ptr) { val = ((u32) (*(u8 *) ptr+1)<< 8) | *(u8 *) ptr; }
      |                                                       ^~~
evg/raster_yuv.c:1537:5: note: in expansion of macro ‘get_u16_le’
 1537 |     get_u16_le(cdst, pU);
      |     ^~~~~~~~~~
evg/raster_yuv.c: In function ‘evg_nv12_10_flush_uv_const’:
evg/raster_yuv.c:1294:55: error: ‘ptr’ undeclared (first use in this function)
 1294 | #define get_u16_le(val, _ptr) { val = ((u32) (*(u8 *) ptr+1)<< 8) | *(u8 *) ptr; }
      |                                                       ^~~
evg/raster_yuv.c:1727:5: note: in expansion of macro ‘get_u16_le’
 1727 |     get_u16_le(dst, pU);
      |     ^~~~~~~~~~
evg/raster_yuv.c: In function ‘evg_nv12_10_flush_uv_var’:
evg/raster_yuv.c:1294:55: error: ‘ptr’ undeclared (first use in this function)
 1294 | #define get_u16_le(val, _ptr) { val = ((u32) (*(u8 *) ptr+1)<< 8) | *(u8 *) ptr; }
      |                                                       ^~~
evg/raster_yuv.c:1769:5: note: in expansion of macro ‘get_u16_le’
 1769 |     get_u16_le(cdst, pU);
      |     ^~~~~~~~~~
evg/raster_yuv.c: In function ‘evg_yuv422p_10_flush_uv_const’:
evg/raster_yuv.c:1294:55: error: ‘ptr’ undeclared (first use in this function)
 1294 | #define get_u16_le(val, _ptr) { val = ((u32) (*(u8 *) ptr+1)<< 8) | *(u8 *) ptr; }
      |                                                       ^~~
evg/raster_yuv.c:1907:5: note: in expansion of macro ‘get_u16_le’
 1907 |     get_u16_le(dst, s_ptr_u);
      |     ^~~~~~~~~~
evg/raster_yuv.c: In function ‘evg_yuv422p_10_flush_uv_var’:
evg/raster_yuv.c:1294:55: error: ‘ptr’ undeclared (first use in this function)
 1294 | #define get_u16_le(val, _ptr) { val = ((u32) (*(u8 *) ptr+1)<< 8) | *(u8 *) ptr; }
      |                                                       ^~~
evg/raster_yuv.c:1941:5: note: in expansion of macro ‘get_u16_le’
 1941 |     get_u16_le(cdst, pU);
      |     ^~~~~~~~~~
make[2]: *** [../config.mak:159: evg/raster_yuv.o] Error 1
````

**Comment by aureliendavid** [comment]
that's a good catch, thanks for reporting

it should be fixed by the commit above, can you retest and confirm?

**Comment by marillat** [comment]
Build fine now. Thanks.

But I see another issue. Another bug report need to be filled 

```
/usr/bin/ld: filter_core/filter_pck.o: in function `gf_filter_aggregate_packets':
./src/filter_core/filter_pck.c:642: undefined reference to `__sync_add_and_fetch_8'
/usr/bin/ld: ./src/filter_core/filter_pck.c:558: undefined reference to `__sync_add_and_fetch_8'
/usr/bin/ld: filter_core/filter_pck.o: in function `gf_filter_pck_send_internal':
./src/filter_core/filter_pck.c:1115: undefined reference to `__sync_add_and_fetch_8'
/usr/bin/ld: ./src/filter_core/filter_pck.c:1037: undefined reference to `__sync_add_and_fetch_8'
/usr/bin/ld: filter_core/filter_pid.o: in function `gf_filter_pid_drop_packet':
./src/filter_core/filter_pid.c:5122: undefined reference to `__sync_sub_and_fetch_8'
/usr/bin/ld: ./src/filter_core/filter_pid.c:5113: undefined reference to `__sync_sub_and_fetch_8'
collect2: error: ld returned 1 exit status
make: *** [Makefile:346: ../bin/gcc/libgpac.so] Error 1
```

**Comment by aureliendavid** [comment]
> But I see another issue. Another bug report need to be filled ?

I think we can stay here. 

This seems a little bit more tricky, looks like powerpc is missing some atomics. 

Do you know if your machine is 32 or 64 bits? 

I'm going to need to find a ppc machine to do some tests.

**Comment by marillat** [comment]
32 bits. I also added  -latomic in LDFLAGS to avoid this error

````
/usr/bin/ld: quickjs/quickjs.o: in function `js_atomics_op':
./src/quickjs/quickjs.c:51410: undefined reference to `__atomic_fetch_add_8'
/usr/bin/ld: ./src/quickjs/quickjs.c:51411: undefined reference to `__atomic_fetch_and_8'
/usr/bin/ld: ./src/quickjs/quickjs.c:51412: undefined reference to `__atomic_fetch_or_8'
/usr/bin/ld: ./src/quickjs/quickjs.c:51429: undefined reference to `__atomic_load_8'
/usr/bin/ld: ./src/quickjs/quickjs.c:51413: undefined reference to `__atomic_fetch_sub_8'
/usr/bin/ld: ./src/quickjs/quickjs.c:51414: undefined reference to `__atomic_fetch_xor_8'
/usr/bin/ld: ./src/quickjs/quickjs.c:51415: undefined reference to `__atomic_exchange_8'
/usr/bin/ld: ./src/quickjs/quickjs.c:51458: undefined reference to `__atomic_compare_exchange_8'
/usr/bin/ld: quickjs/quickjs.o: in function `js_atomics_store':
./src/quickjs/quickjs.c:51522: undefined reference to `__atomic_store_8'
collect2: error: ld returned 1 exit status
````

**Comment by aureliendavid** [comment]
to clarify: -latomic resolves the second error (on quickjs) and not the first (on filter_core) right?

**Comment by jeanlf** [comment]
can you try in src/quickjs/quickjs.c to comment (line 79) 
`#define CONFIG_ATOMICS`
 ?

**Comment by marillat** [comment]
> to clarify: -latomic resolves the second error (on quickjs) and not the first (on filter_core) right?

Yes.

> can you try in src/quickjs/quickjs.c to comment (line 79)
> `#define CONFIG_ATOMICS`
> ?
Still the same.

**Comment by aureliendavid** [comment]
ok 

https://gcc.gnu.org/onlinedocs/gcc-4.1.2/gcc/Atomic-Builtins.html says:

_"Not all operations are supported by all target processors. If a particular operation cannot be implemented on the target processor, a warning will be generated and a call an external function will be generated. The external function will carry the same name as the builtin, with an additional suffix `_n' where n is the size of the data type. "_

since you have a 32bits arch and only functions with `_8` are reported missing I'm guessing ppc 32bits doesn't have the builtins for 64 bits ints

but since you fixed the quickjs error by linking with libatomic, I'm guessing that for ppc we can use __atomic in libgpac and force linking with libatomic 

I'm still gonna try to get a vm running to test this. What's the debian arch you're using exactly?

(If you want to test it yourself I'd try editing `include/gpac/thread.h` and replacing:

```c
#define safe_int64_add(__v, inc_val) __sync_add_and_fetch((int64_t *) (__v), inc_val)
```

with 

```c
#define safe_int64_add(__v, inc_val) __atomic_add_fetch((int64_t *) (__v), inc_val,  __ATOMIC_ACQ_REL)
```

but I haven't tested this at all yet)

**Comment by marillat** [comment]
> I'm still gonna try to get a vm running to test this. What's the debian arch you're using exactly?

64 bits kernel with 32 bits packages on Debian unstable from Debian ports.

> (If you want to test it yourself I'd try editing `include/gpac/thread.h` and replacing:
> 
> ```c
> #define safe_int64_add(__v, inc_val) __sync_add_and_fetch((int64_t *) (__v), inc_val)
> ```
> 
> with
> 
> ```c
> #define safe_int64_add(__v, inc_val) __atomic_add_fetch((int64_t *) (__v), inc_val,  __ATOMIC_ACQ_REL)
> ```

Still the same.
If you want I can ask on the debian-ppc mailinglist ?
````
/usr/bin/ld: filter_core/filter_pck.o: in function `gf_filter_aggregate_packets':
./src/filter_core/filter_pck.c:642: undefined reference to `__atomic_fetch_add_8'
/usr/bin/ld: ./src/filter_core/filter_pck.c:558: undefined reference to `__atomic_fetch_add_8'
/usr/bin/ld: filter_core/filter_pck.o: in function `gf_filter_pck_send_internal':
./src/filter_core/filter_pck.c:1115: undefined reference to `__atomic_fetch_add_8'
/usr/bin/ld: ./src/filter_core/filter_pck.c:1037: undefined reference to `__atomic_fetch_add_8'
/usr/bin/ld: filter_core/filter_pid.o: in function `gf_filter_pid_drop_packet':
./src/filter_core/filter_pid.c:5122: undefined reference to `__sync_sub_and_fetch_8'
/usr/bin/ld: ./src/filter_core/filter_pid.c:5113: undefined reference to `__sync_sub_and_fetch_8'
/usr/bin/ld: quickjs/quickjs.o: in function `js_atomics_op':
./src/quickjs/quickjs.c:51410: undefined reference to `__atomic_fetch_add_8'
/usr/bin/ld: ./src/quickjs/quickjs.c:51411: undefined reference to `__atomic_fetch_and_8'
/usr/bin/ld: ./src/quickjs/quickjs.c:51412: undefined reference to `__atomic_fetch_or_8'
/usr/bin/ld: ./src/quickjs/quickjs.c:51429: undefined reference to `__atomic_load_8'
/usr/bin/ld: ./src/quickjs/quickjs.c:51413: undefined reference to `__atomic_fetch_sub_8'
/usr/bin/ld: ./src/quickjs/quickjs.c:51414: undefined reference to `__atomic_fetch_xor_8'
/usr/bin/ld: ./src/quickjs/quickjs.c:51415: undefined reference to `__atomic_exchange_8'
/usr/bin/ld: ./src/quickjs/quickjs.c:51458: undefined reference to `__atomic_compare_exchange_8'
/usr/bin/ld: quickjs/quickjs.o: in function `js_atomics_store':
./src/quickjs/quickjs.c:51522: undefined reference to `__atomic_store_8'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:346: ../bin/gcc/libgpac.so] Error 1
make[2]: Leaving directory '/home/marillat/gpac-dmo-1.0.1/src'
make[1]: *** [Makefile:13: all] Error 2
make[1]: Leaving directory '/home/marillat/gpac-dmo-1.0.1'
dh_auto_build: error: make -j5 returned exit code 2
````

**Comment by marillat** [comment]
I forgot to add -laromic to LDFLAGS
I see only this small error now.
````
usr/bin/ld: filter_core/filter_pid.o: in function `gf_filter_pid_drop_packet':
./src/filter_core/filter_pid.c:5122: undefined reference to `__sync_sub_and_fetch_8'
/usr/bin/ld: ./src/filter_core/filter_pid.c:5113: undefined reference to `__sync_sub_and_fetch_8'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:346: ../bin/gcc/libgpac.so] Error 1
````

**Comment by aureliendavid** [comment]
same, in `thread.h` replace 

`#define safe_int64_sub(__v, dec_val) __sync_sub_and_fetch((int64_t *) (__v), dec_val)`

with 

`#define safe_int64_sub(__v, dec_val) __atomic_sub_fetch((int64_t *) (__v), dec_val, __ATOMIC_ACQ_REL)`

with -latomic still

**Comment by marillat** [comment]
> same, in `thread.h` replace
> 
> `#define safe_int64_sub(__v, dec_val) __sync_sub_and_fetch((int64_t *) (__v), dec_val)`
> 
> with
> 
> `#define safe_int64_sub(__v, dec_val) __atomic_sub_fetch((int64_t *) (__v), dec_val, __ATOMIC_ACQ_REL)`
> 
> with -latomic still

That fix this bug.

I did a build for amd64 and i386 without problem.

Thanks again.
Christian

**Comment by aureliendavid** [comment]
> > I'm still gonna try to get a vm running to test this. What's the debian arch you're using exactly?
> 
> 64 bits kernel with 32 bits packages on Debian unstable from Debian ports.

I'm struggling a bit to replicate this setup with vms. Could you post your:

`uname -a`
`lsb_release -a`
`cat /etc/apt/sources.list`

? Thanks

**Comment by marillat** [comment]
> `uname -a`

Linux fabian.marillat.net 5.6.0-2-powerpc64 #1 SMP Debian 5.6.14-2 (2020-06-09) ppc64 GNU/Linux
> `lsb_release -a`

No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux bullseye/sid
Release:	unstable
Codename:	sid
> `cat /etc/apt/sources.list`

deb https://www.deb-multimedia.org/ sid main
deb http://ftp.ports.debian.org/debian-ports/ unstable main
deb http://ftp.ports.debian.org/debian-ports/ unreleased main

**Comment by aureliendavid** [comment]
I finally managed to (painfully) set up a powerpc debian sid vm and reproduce the issue. 

It takes about 2 hours to build a minimal libgpac (!), so the patch is coming... slowly

**Comment by aureliendavid** [comment]
this should now be fixed

@marillat can you retest with bb28ed22a123f81369c4153949df40f454eb20c7 and report back?

thanks

**Comment by marillat** [comment]
@aureliendavid  build fine for amd64, armhf, i386 and powerpc architectures.

Thanks

**Comment by aureliendavid** [comment]
nice, thanks!
